### ISSUE TITLE
how to write a gif making fro multiple images

### ISSUE THREAD

**Issue by harper1999** [imageio/imageio#829]
previously, imageio.mimsave has the param frames receiving a list. Not availbale for v3.imwrite anymore

**Comment by FirefoxMetzger** [comment]
@harper1999 Could you share a small example to demonstrate the problem?

The following works fine on my machine:

```python
>>> import imageio.v3 as iio
>>> frames = iio.imread("imageio:newtonscradle.gif")
>>> iio.imwrite("foo.gif", frames)
>>> iio.imwrite("foo.gif", frames, loop=0) 
>>> frame_list = [x for x in frames]
>>> iio.imwrite("foo2.gif", frame_list, loop=0) 
>>> type(frame_list)
<class 'list'>
```

It produces two GIFs as expected:

![foo](https://user-images.githubusercontent.com/4402489/177612402-246e2bd0-5e41-4f34-816b-19d0692f23c7.gif)

![foo2](https://user-images.githubusercontent.com/4402489/177612416-d6fa5392-d336-4fb5-9a0d-0015c37cff99.gif)

**Comment by FirefoxMetzger** [comment]
@harper1999 Did that answer your question? If so, I'd suggest we close this issue; if not, could you describe your problem in more detail?

**Comment by FirefoxMetzger** [comment]
I will close this question due to inactivity. However, if the problem persists feel free to leave a note and I'm happy to reopen :)

**Comment by harper1999** [comment]
# trying to make all pngs into 1 gif
 frames = [imageio.v3.imread(file) for file in files]  #  multiple pngs in the files
    imageio.mimsave(gif_name, frames, format='GIF-PIL', duration=2)

how does iio.imwrite work in this situation?

**Comment by FirefoxMetzger** [comment]
I would use `imageio.v3.imwrite` for this, because it can handle a list of frames. `imageio.imwrite` is (in ImageIO 2.x) an alias for `imageio.v2.imwrite` which can only handle a single frame. Using the v3 API you would do:

```python
import imageio.v3 as iio

frames = [iio.imread(file) for file in files]
iio.imwrite(
    gif_name,  # name of the resulting file, e.g. fancy.gif
    frames,  # the images that make up the gif
    duration=2, #  show each image for 2 seconds
    loop=0,  # number of repeats, loop=0 means repeat forever
)
```

Note that all images/frames have to have the same shape. If not, you will get an error that the shapes are mismatched which GIF does not support.

**Comment by harper1999** [comment]
I tried it this way, but not lucky. On the frames, it pops up Expected type 'ndarray', got 'list' instead

**Comment by FirefoxMetzger** [comment]
Hm, which version of ImageIO are you using? (`python -c "import imageio; print(imageio.__version__)"` if it isn't 2.19.5 you could try upgrading and checking if the problem goes away.

If that doesn't solve it there are two additional things we could try:

1. Are you able to share the images? If so, I can see if it works on my computer and if not investigate where the problem comes from.
2. Could you check if the commands I shared earlier (https://github.com/imageio/imageio/issues/829#issuecomment-1176510972) work as expected?

**Comment by harper1999** [comment]
after update to 2.19.5 still not working. Commands don't work either
![ddd](https://user-images.githubusercontent.com/53476659/180793916-b226c254-ae46-4fa1-b393-8960e4b1d767.jpg)

**Comment by harper1999** [comment]
my code is below for your reference.  I have one background image in source_folder, and some images in the source_folder/new folder 

`import imageio
import imghdr
import os
from PIL import Image

source_folder = os.path.dirname(os.path.realpath(__file__)) + r'\GIF_Source'
img_types = {'png', 'jpg', 'jpeg', 'tif', 'bmp', 'gif'}


def paste_background(sub_dir, dirname, filenames):
    images = [Image.open(sub_dir + r'\\' + file).convert(mode='RGBA') for file in filenames if dirname in file]
    widths = [image.size[0] for image in images]  # height
    heights = [image.size[1] for image in images]  # width
    max_width = max(widths)
    max_height = max(heights)
    max_value = max(max_width, max_height)

    background = Image.open(source_folder + r'\background.jpg').convert(mode='RGB')

    new_names = []
    for i in range(len(images)):
        r_background = background.resize((max_value, max_value))
        width_offset = int((max_value - widths[i]) / 2)
        height_offset = int((max_value - heights[i]) / 2)
        r_background.paste(images[i], (width_offset, height_offset), mask=images[i])  # int默认向下取整
        r_background.save(sub_dir + r'\\' + str(i + 1) + '.jpg', quality=95)
        new_names.append(i + 1)


def compose_gif(sub_dir, filenames):
    filtered_files = [file for file in filenames if
                      file.split('.')[0].isdigit() and imghdr.what(os.path.join(sub_dir, file)) != 'gif']
    frames = [imageio.v3.imread(sub_dir + r'\\' + file) for file in filtered_files]
    gif_name = str(int(filtered_files[-1].split('.')[0]) + 1) + '.gif'
    # imageio.mimsave(sub_dir + r'\\' + gif_name, frames, format='GIF-PIL', duration=2)  # this one works
    imageio.v3.imwrite(sub_dir + r'\\' + gif_name, frames, duration=2)


def main():
    for root, dirnames, files in os.walk(source_folder):
        for dirname in dirnames:
            sub_dir = os.path.join(root, dirname)
            filenames = os.listdir(sub_dir)
            # check if the image types are supported
            if filenames and (imghdr.what(str(os.path.join(sub_dir, file))) in img_types for file in filenames):
                paste_background(sub_dir, dirname, filenames)
                compose_gif(sub_dir, filenames)


if __name__ == '__main__':
    main()`

**Comment by FirefoxMetzger** [comment]
Okay, I tested the code that you shared. Unfortunately, I can't replicate the issue, because it is working fine on the images that I've tried it on. 

Here is the snippet that I used to generate the background image and some frames for the test:

```python
# get_sample.py
import imageio.v3 as iio
from pathlib import Path

root = Path("GIF_Source")
root.mkdir(exist_ok=True)

background = iio.imread("https://img.freepik.com/free-photo/top-view-background-beautiful-white-grey-brown-cream-blue-background_140725-72219.jpg?w=2000")
iio.imwrite("GIF_Source/background.jpg", background)

(root / "new").mkdir(exist_ok=True)
frames = iio.imread("https://i.imgur.com/uNdDoPT.gif")
for idx, frame in enumerate(frames):
    iio.imwrite(f"GIF_Source/new/{idx+1}.jpg", frame)

```

I then had to make one change to your code in order to make it run to the point where it would generate a gif:

```diff
@@ -8,7 +8,7 @@ img_types = {'png', 'jpg', 'jpeg', 'tif', 'bmp', 'gif'}


 def paste_background(sub_dir, dirname, filenames):
-    images = [Image.open(sub_dir + r'\\' + file).convert(mode='RGBA') for file in filenames if dirname in file]
+    images = [Image.open(sub_dir + r'\\' + file).convert(mode='RGBA') for file in filenames]
     widths = [image.size[0] for image in images]  # height
     heights = [image.size[1] for image in images]  # width
     max_width = max(widths)
```

After this, the code ran without issues and produced the following result (resized to fit GitHubs filesize limit):

![foo](https://user-images.githubusercontent.com/4402489/180811456-d4f121cd-d872-4baa-af17-8b9f24e20248.gif)


This shows that, for the images I chose, `iio.v3.imwrite` does produce a GIF as you intended. I suspect there is a problem specific to your images or to the way your filenames interact with your program.

---

On a different note, you actually don't need to temporarily save/write the images back to disk. Instead, you can convert them  into an ndarray directly by calling `np.ndarray(r_background)`. This saves you some trouble with preserving the order of frames when you have many and it also makes your code cleaner. For example the following:

```python
# get_sample.py

import imageio.v3 as iio
from pathlib import Path

root = Path("GIF_Source")
root.mkdir(exist_ok=True)

background = iio.imread("https://img.freepik.com/free-photo/top-view-background-beautiful-white-grey-brown-cream-blue-background_140725-72219.jpg?w=2000")
iio.imwrite("GIF_Source/background.jpg", background)

(root / "new").mkdir(exist_ok=True)
frames = iio.imread("https://i.imgur.com/uNdDoPT.gif", mode="RGBA")
for idx, frame in enumerate(frames):
    iio.imwrite(f"GIF_Source/new/frame_{idx:03d}.png", frame)

```

```python
# refactored.py

import imageio.v3 as iio
from PIL import Image
from pathlib import Path
import numpy as np

source_folder = Path(__file__).parent / "GIF_Source"
dest_folder = Path(__file__).parent


def paste_background(directory:Path, background_image:Image):
    files = [file for file in directory.iterdir() if file.is_file()]
    images = [Image.open(x).convert(mode="RGBA") for x in files]

    widths = [image.size[0] for image in images]
    heights = [image.size[1] for image in images]

    max_width = max(widths)
    max_height = max(heights)
    max_value = max(max_width, max_height)

    frames = list()
    for image in images:
        r_background = background_image.resize((max_value, max_value))
        width_offset = int((max_value - image.size[0]) / 2)
        height_offset = int((max_value - image.size[1]) / 2)
        r_background.paste(image, (width_offset, height_offset), mask=image)
        frames.append(np.asarray(r_background))
    
    return frames


def main():
    background_image = Image.open(source_folder / 'background.jpg').convert(mode='RGB')

    for directory in source_folder.iterdir():
        if not directory.is_dir():
            continue

        frames = paste_background(directory, background_image)

        gif_name = directory.name + ".gif"
        iio.imwrite(dest_folder / gif_name, frames, loop=0, duration=40)


if __name__ == '__main__':
    main()

```

Produces a GIF that is much close to what you'd expect (again resized to fit GHs size limit):
![small](https://user-images.githubusercontent.com/4402489/180822354-2d8b0138-808f-4718-b9bc-d35509563cae.gif)

**Comment by harper1999** [comment]
Thanks a lot for your help. I'm using Pycharm. Unfortunately, with the type hint, it raises an exception if I run it. Then I tried to convert the images pasted on background into ndarray and call imwrite, it went through, but it came up with a new problem. That is the param duration, it looks like the unit of it is no longer second, but millisecond (after testing to set it 1000). I wonder why.

Then I tried your refactored code and it showed same warning hint 'unexpected types'.  But it's ok to run it through despite the hint this time. But again, the duration has to be set like 1000. Would you please send me the source images you tried to test it ?

**Comment by FirefoxMetzger** [comment]
> That is the param duration, it looks like the unit of it is no longer second, but millisecond (after testing to set it 1000). I wonder why.

If you call `imageio.v3.imwrite` then you will use the new pillow plugin by default (faster and more features), whereas `imageio.v2.imwrite` will use the old/legacy plugin. Pillow (which is our default backend for writing GIF) uses a duration in milliseconds when writing a GIF and we decided to match this in the V3 pillow plugin because it means that there is less "magic" happening between a user making a request and the backend doing the work.

> Then I tried your refactored code and it showed same warning hint 'unexpected types'.

You mean that you still see the PyCharm hint that you are inserting a `list[ndarray]` where it expects a `ndarray`? If so, I've created a PR to fix this (#848) and that should get fixed with the next release on Monday.

> Would you please send me the source images you tried to test it ?

You can download the images yourself. In the comment above I shared the script that will download the files and set up the `GIF_Source` folder:

```python
# get_sample.py

import imageio.v3 as iio
from pathlib import Path

root = Path("GIF_Source")
root.mkdir(exist_ok=True)

background = iio.imread("https://img.freepik.com/free-photo/top-view-background-beautiful-white-grey-brown-cream-blue-background_140725-72219.jpg?w=2000")
iio.imwrite("GIF_Source/background.jpg", background)

(root / "new").mkdir(exist_ok=True)
frames = iio.imread("https://i.imgur.com/uNdDoPT.gif", mode="RGBA")
for idx, frame in enumerate(frames):
    iio.imwrite(f"GIF_Source/new/frame_{idx:03d}.png", frame)
```
